crda/remotematcher.go

```diff
@@ -0,0 +1,389 @@
+	Key         string `json:"key" yaml:"key"`
+}
+
+// Configure implements driver.MatcherConfigurable.
+func (m *Matcher) Configure(ctx context.Context, f driver.MatcherConfigUnmarshaler, c *http.Client) error {
+	var cfg Config
+	if err := f(&cfg); err != nil {
+		return err
+	}
+
+	if cfg.Concurrency > 0 {
+		m.requestConcurrency = cfg.Concurrency
+	}
+
+	if cfg.URL != "" {
+		u, err := url.Parse(cfg.URL)
+		if err != nil {
+			return err
+		}
+		m.url = u
+	}
+	if cfg.Source != "" {
+		m.source = cfg.Source
+	}
+	if cfg.Key != "" {
+		m.key = cfg.Key
+	}
+	m.client = c
+
+	return nil
+}
+
+// QueryRemoteMatcher implements driver.RemoteMatcher.
+func (m *Matcher) QueryRemoteMatcher(ctx context.Context, records []*claircore.IndexRecord) (map[string][]*claircore.Vulnerability, error) {
+	ctx = baggage.ContextWithValues(ctx,
+		label.String("component", "crda/Matcher.QueryRemoteMatcher"))
+	zlog.Debug(ctx).
+		Int("records", len(records)).
+		Msg("request")
+
+	// Map Package{name@version} to Package to associate it with Vulnerability.
+	packageVersionToIndexRecord := make(map[string]*claircore.IndexRecord)
+	key := func(Name, Version string) string {
+		return fmt.Sprintf("%s@%s", Name, Version)
+	}
+	for _, ir := range records {
+		packageVersionToIndexRecord[key(ir.Package.Name, ir.Package.Version)] = ir
+	}
+
+	results := make(map[string][]*claircore.Vulnerability)
+	ctrlC, errorC := m.invokeComponentAnalysesInBatch(ctx, records)
+	err := <-errorC // guaranteed to have an err or be closed.
+	// Don't propagate error, log and move on.
+	if err != nil {
+		zlog.Error(ctx).Err(err).Msg("remote api call failure")
+		return results, nil
+	}
+	for vrs := range ctrlC {
+		for _, vr := range vrs {
+			ir := packageVersionToIndexRecord[key(vr.Name, vr.Version)]
+
+			// A package can have 0 or more vulnerabilities for a version.
+			var vulns []*claircore.Vulnerability
+			for _, vuln := range vr.Vulnerabilities {
+				vulns = append(vulns, &claircore.Vulnerability{
+					ID:                 vuln.ID,
+					Updater:            "CodeReadyAnalytics",
+					Name:               vuln.ID,
+					Description:        vuln.Title,
+					Links:              vuln.URL,
+					Severity:           vuln.Severity,
+					NormalizedSeverity: NormalizeSeverity(vuln.Severity),
+					FixedInVersion:     strings.Join(vuln.FixedIn, ", "),
+					Package:            ir.Package,
+					Repo:               ir.Repository,
+				})
+			}
+			results[ir.Package.ID] = append(results[ir.Package.ID], vulns...)
+		}
+	}
+	zlog.Debug(ctx).
+		Int("vulnerabilities", len(results)).
+		Msg("response")
+	return results, nil
+}
+
+func (m *Matcher) invokeComponentAnalysesInBatch(ctx context.Context, records []*claircore.IndexRecord) (<-chan []*VulnReport, <-chan error) {
+	inC := make(chan []*claircore.IndexRecord, m.requestConcurrency)
+	ctrlC := make(chan []*VulnReport, m.requestConcurrency)
+	errorC := make(chan error, 1)
+	batchSize := m.batchSize
+	go func() {
```

Comment:
I may be wrong, but I think this could be unnecessarily complicated. This goroutine exists to allow the function to return before the work is done, but the very next line basically blocks until work is done:
```go
	ctrlC, errorC := m.invokeComponentAnalysesInBatch(ctx, records)
	err := <-errorC // guaranteed to have an err or be closed.
``` 
What would we lose from removing this goroutine and returning `([]*VulnReport, error)`? (To be clear, I still think the errgroup is good).

Reply:
I think the idea is to make the api calls parallelly.Even thought it waits for the error.
I will @arajkumar To explain this better as he had implemented this. :)

Reply:
Gotcha and I agree, the API calls in parallel makes sense, I'm just suggesting that the synchronization takes place in the function rather than the caller needing to be responsible for that.

Reply:
Resolved.